artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/BucketMessageGroups.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.core.server.impl;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Predicate;
+import org.apache.activemq.artemis.api.core.SimpleString;
+
+/**
+ * BucketMessageGroups, stores values against a bucket, where the bucket used is based on the provided key objects hash.
+ *
+ * As such where keys compute to the same bucket they will act on that stored value, not the unique specific key.
+ *
+ * The number of buckets is provided at construction.
+ */
+public class BucketMessageGroups<C> implements MessageGroups<C> {
+
+   //This _AMQ_GROUP_BUCKET_INT_KEY uses the post-fixed value after this key, as an int, it is used for a few cases:
+   //1) For the admin screen we need to show a group key so we have to map back from int to something, as it expects SimpleString.
+   //2) Admin users still need to interact with a specific bucket/group e.g. they may need to reset a bucket.
+   //3) Choice of key is we want to avoid risk of clashing with users groups keys.
+   //4) Actually makes testing a little easier as we know how the parsed int will hash.
+   private static SimpleString _AMQ_GROUP_BUCKET_INT_KEY = new SimpleString("_AMQ_GROUP_BUCKET_INT_KEY_");
+
+   private final int bucketCount;
+   private C[] buckets;
+   private int size = 0;
+
+   public BucketMessageGroups(int bucketCount) {
+      if (bucketCount < 1) {
+         throw new IllegalArgumentException("Bucket count must be greater than 0");
+      }
+      this.bucketCount = bucketCount;
+   }
+
+   private int getBucket(SimpleString key) {
+      Object bucketKey = key;
+      if (key.startsWith(_AMQ_GROUP_BUCKET_INT_KEY)) {
+         bucketKey = retrieveBucketIntFromKey(key);
+      }
+      return getHashBucket(bucketKey, bucketCount);
+   }
+
+   private static int getHashBucket(final Object key, final int bucketCount) {
+      return (key.hashCode() & Integer.MAX_VALUE) % bucketCount;
```

Comment:
% is a quite heavy operation: masking with a power of 2 bucketCount would be better if possible.
This is suppose to be on the hot path?

Reply:
Ive responded to this same comment several times yes. That only works if buckets is power of two and we have to honour the bucket count set.

The actual performance compared to everything is tiny. Note we actually have mod in journal code fyi.

Reply:
Did you implement this from scratch or did you base it on the ActiveMQ5 bucket?

We had an outstanding task to implement it based on AMQ5 version.

Reply:
Ohh.. you hash the GroupID... got it!

I feel dumb that I don't understand why the bitwise operation.

Reply:
that's just stripping out negatives.. got it!

Reply:
@jbertram pointed me this is the same on AMQ5

Reply:
Yes its similar to amq5.